// react-device-identifier/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var DeviceIdentifier =
/*#__PURE__*/
function (_React$PureComponent) {
  _inherits(DeviceIdentifier, _React$PureComponent);

  function DeviceIdentifier() {
    var _this;

    _classCallCheck(this, DeviceIdentifier);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(DeviceIdentifier).call(this));
    _this.state = {
      isMobile: false,
      isDesktop: false,
      isTablet: false // can also check for tablets

    };
    return _this;
  }

  _createClass(DeviceIdentifier, [{
    key: "handleWindowResize",
    value: function handleWindowResize() {
      var resolution = window.innerWidth;
      var isMobile = resolution >= 320 && resolution <= 480;
      var isTablet = resolution >= 768 && resolution <= 1024;
      var isDesktop = !isMobile && !isTablet;
      this.setState({
        isMobile: isMobile,
        isTablet: isTablet,
        isDesktop: isDesktop
      });
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.handleWindowResize();
      window.addEventListener("resize", this.handleWindowResize.bind(this));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          isMobile = _this$state.isMobile,
          isTablet = _this$state.isTablet,
          isDesktop = _this$state.isDesktop;
      var _this$props = this.props,
          _this$props$isDesktop = _this$props.isDesktop,
          propsIsDesktop = _this$props$isDesktop === void 0 ? false : _this$props$isDesktop,
          _this$props$isMobile = _this$props.isMobile,
          propsIsMobile = _this$props$isMobile === void 0 ? false : _this$props$isMobile,
          _this$props$isTablet = _this$props.isTablet,
          propsIsTablet = _this$props$isTablet === void 0 ? false : _this$props$isTablet;

      if (isDesktop && propsIsDesktop || isMobile && propsIsMobile || isTablet && propsIsTablet // can add more checks like tablets
      ) {
          return _react["default"].createElement(_react["default"].Fragment, null, this.props.children);
        }

      return null;
    }
  }]);

  return DeviceIdentifier;
}(_react["default"].PureComponent);

exports["default"] = DeviceIdentifier;
