// @xdn/core/router/validateWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * List of header regexes that will be matched against user provided header name.
 * The following headers are not allowed for user to modify in any way.
 *
 * * `x-xdn-*` headers are generated and maintained by XDN platform to store intermediate state during request processing
 * * `host` header should only be manipulated by the XDN platform as otherwise it messes with reverse proxying
 * * `x-request-id` header is generated by XDN platform and used as unique identifier of requests
 * * `content-length` header cannot be manipulated apart from manipulating the body and we don't support that in the edge
 */
var XDN_PROHIBITED_HEADER_NAMES = [/^x-xdn-/i, /^host$/i, /^x-request-id$/i, /^content-length$/i];
/**
 * List of cookie regexes that will be matched against user provided cookie name.
 * The following cookies are not allowed for user to modify in any way.
 *
 * * `xdn_*` cookies are created and maintained by XDN platform (e.g. `xdn_buckets` and `xdn_destination`)
 */
var XDN_PROHIBITED_COOKIE_NAMES = [/^xdn_/i];
/**
 * Validate that prohibited header names are not overrideable by the user
 * @param name
 */
var validateProhibitedHeaderNames = function (name) {
    if (XDN_PROHIBITED_HEADER_NAMES.find(function (prohibitedRegex) { return name.match(prohibitedRegex); })) {
        throw new Error("Header \"" + name + "\" is reserved and cannot be modified");
    }
};
/**
 * Validate that prohibited cookie names are not overrideable by the user
 * @param name
 */
var validateProhibitedCookieNames = function (name) {
    if (XDN_PROHIBITED_COOKIE_NAMES.find(function (prohibitedRegex) { return name.match(prohibitedRegex); })) {
        throw new Error("Cookie \"" + name + "\" is reserved and cannot be modified");
    }
};
/**
 * Validate Set response header
 * @param name header name
 * @param value header value
 */
exports.validateSetResponseHeader = function (name /* value */) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Update response header
 * @param name header name
 * @param match match pattern
 * @param replace replace value
 */
exports.validateUpdateResponseHeader = function (name /* match, replace */) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Remove response header
 * @param name
 */
exports.validateRemoveResponseHeader = function (name) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate set request header
 * @param name header name
 * @param value header value
 */
exports.validateSetRequestHeader = function (name) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Update request header
 * @param name header name
 * @param match match pattern
 * @param replace replace value
 */
exports.validateUpdateRequestHeader = function (name) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Remove request header
 * @param name
 */
exports.validateRemoveRequestHeader = function (name) {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate cookie name.
 * @param name
 */
exports.validateCookieName = function (name) {
    validateProhibitedCookieNames(name);
};
