// @xdn/cli/commands/run.js
"use strict";

const {
  join
} = require('path');

const {
  existsSync
} = require('fs');

const chalk = require('chalk');

const logo = require('../utils/logo');

const runDeploymentArchive = require('../utils/runDeploymentArchive');

const runWithServerless = require('../utils/runWithServerless');

const clearPorts = require('../utils/clearPorts');

const getEntryPoint = require('../utils/getEntryPoint');

exports.command = 'run [archive]';
exports.describe = 'Runs your project locally, simulating the XDN cloud environment. When no arguments are provided, this command is the same as xdn dev.';

exports.builder = yargs => yargs.option('production', {
  type: 'boolean',
  alias: 'p',
  describe: 'Runs your app in production mode, with caching enabled, emulating the XDN serverless runtime environment. You must first run xdn build to create a production build of your app.'
}).option('cache', {
  type: 'boolean',
  alias: 'c',
  describe: 'Enables caching in development mode. Caching is enabled by default in production mode.'
}).positional('archive', {
  describe: 'The path to a deployment archive (zip) file downloaded from the XDN Developer Console'
});

exports.handler = async ({
  context,
  production,
  archive,
  cache
}) => {
  await clearPorts();
  const xdnDir = join(process.cwd(), '.xdn');
  const {
    logger
  } = context;

  if (process.env.NODE_ENV === 'production') {
    production = true;
    cache = true;
  }

  if (archive) {
    logger.info(`> Running deployed app locally using ${logo}...`);
    await runDeploymentArchive(archive);
  } else if (production) {
    if (!existsSync(xdnDir)) {
      logger.error(`\nXDN production build not found. Please run ${chalk.green('xdn build')} before running ${chalk.green('xdn run --production')}.\n`);
      process.exit(1);
    }

    logger.info(`> Running ${logo} in production mode...`);
    await runWithServerless(xdnDir);
  } else {
    logger.info(`> Starting ${logo} in development mode with caching ${cache ? chalk.green('enabled') : chalk.red('disabled')}...`);

    if (cache) {
      process.env.XDN_CACHE = 'true';
    }

    const dev = await getEntryPoint('dev');
    dev();
  }
};
